pkg/cloudevents/types/value.go

```diff
@@ -0,0 +1,210 @@
+	case float32, float64:
+		f := reflect.ValueOf(x).Float()
+		if f > math.MaxInt32 || f < math.MinInt32 {
+			return Value{}, rangeErr(x)
+		}
+		return Value{int32(f)}, nil
+
+	case URIRef:
+		return Value{x.URL}, nil
+	case URI:
+		return Value{x.URL}, nil
+	case URLRef:
+		return Value{x.URL}, nil
+
+	case Timestamp:
+		return Value{x.Time}, nil
+
+	case *url.URL, *URIRef, *URLRef, *URI, *Timestamp, *time.Time:
+		rx := reflect.ValueOf(x)
+		if rx.IsNil() {
+			return Value{}, valueErr("", x)
+		}
+		return ValueOf(rx.Elem().Interface())
+
+	default:
+		return Value{}, fmt.Errorf("%T is not a CloudEvents type", x)
+	}
+}
+
+func ValueOfBool(x bool) Value           { return Value{x} }
+func ValueOfInteger(x int32) Value       { return Value{x} }
+func ValueOfString(x string) Value       { return Value{x} }
+func ValueOfBinary(x []byte) Value       { return Value{x} }
+func ValueOfURIRef(x URIRef) Value       { return Value{x} }
+func ValueOfTimestamp(x time.Time) Value { return Value{x} }
+
+// String returns the canonical string encoded representation of the value.
+// Use ToString() to extract the value only if it is of type string.
+func (v Value) String() string {
+	switch x := v.Interface().(type) {
+	case bool:
+		return strconv.FormatBool(x)
+	case int32:
+		return strconv.Itoa(int(x))
+	case string:
+		return x
+	case []byte:
+		return base64.StdEncoding.EncodeToString(x)
+	case url.URL:
+		return x.String()
+	case time.Time:
+		return x.UTC().Format(time.RFC3339Nano)
+	case nil:
+		return "<nil>"
+	default:
+		return fmt.Sprintf("<unknown:%T>", v.v)
+	}
+}
+
+// ToBool returns a bool value, decoding from string if necessary.
+func (v Value) ToBool() (bool, error) {
+	switch x := v.v.(type) {
+	case bool:
+		return x, nil
+	case string:
+		return strconv.ParseBool(x)
+	default:
+		return false, typeErr("Bool", x)
+	}
+}
+
+// ToInteger returns an int32 value, decoding from string if necessary.
+func (v Value) ToInteger() (int32, error) {
+	switch x := v.v.(type) {
+	case int32:
+		return x, nil
+	case string:
+		// Accept floating-point but truncate to int32 as per CE spec.
+		f, err := strconv.ParseFloat(x, 64)
+		if f > math.MaxInt32 || f < math.MinInt32 {
+			return 0, rangeErr(x)
+		}
+		return int32(f), err
+	default:
+		return 0, typeErr("Integer", x)
+	}
+}
+
+// ToString returns a string value. It does NOT encode non-string, use String() for that.
+func (v Value) ToString() (string, error) {
+	switch x := v.v.(type) {
+	case string:
+		return x, nil
+	default:
+		return "", typeErr("String", x)
+	}
+}
+
+// ToBinary returns a []byte value, decoding from string if necessary.
+func (v Value) ToBinary() ([]byte, error) {
+	switch x := v.v.(type) {
+	case []byte:
+		return x, nil
+	case string:
+		return base64.StdEncoding.DecodeString(x)
+	default:
+		return nil, typeErr("Binary", x)
+	}
+}
+
+// ToURIRef returns a url.URL value, decoding from string if necessary.
+func (v Value) ToURIRef() (url.URL, error) {
```

Comment:
If this is converting to `url.URL`, I think it should be `ToURL`. Or make two methods that produce `types.URI` and `types.URIRef`